### PR TITLE
[GOBBLIN-1251] Propagate exception in TaskStateTracker for caller to trigger Helix retry

### PR DESCRIPTION
**Pull request by autumnust**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- https://issues.apache.org/jira/browse/GOBBLIN-1251

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
`registerNewTask` called in `org.apache.gobblin.runtime.GobblinMultiTaskAttempt#runWorkUnits(org.apache.gobblin.runtime.CountUpAndDownLatch)` would swallow the exception if scheduling failed before this change. Post the change, the exception-handling logic within `runWorkunits` will handle it and retry if necessary. 

Also did some comment-enhancement and variable renaming which doesn't break atomicity for this PR since they are relevant.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 1]
```diff
@@ -385,7 +387,10 @@ private boolean taskSuccessfulInPriorAttempt(String taskId) {
   private Pair<List<Task>, Boolean> runWorkUnits(CountUpAndDownLatch countDownLatch) {
 
     List<Task> tasks = Lists.newArrayList();
-    boolean isTaskCreatedSuccessfully = true;
+
+    // A flag indicating if there are any tasks not submitted successfully.
+    // Caller of this method should handler un-submitted task accordingly.
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java:392]: handler -> handle.

un-submitted task -> tasks with submission failures

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 2]
```diff
@@ -426,7 +431,7 @@ private boolean taskSuccessfulInPriorAttempt(String taskId) {
         if (task == null) {
           if (e instanceof RetryException) {
             // Indicating task being null due to failure in creation even after retrying.
-            isTaskCreatedSuccessfully = false;
+            areAllTaskSubmitted = false;
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java:434]: areAllTaskSubmitted -> areAllTasksSubmitted.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java [hunk 3]
```diff
@@ -59,7 +58,11 @@ public void registerNewTask(Task task) {
     try {
       this.scheduledReporters.put(task.getTaskId(), scheduleTaskMetricsUpdater(new TaskMetricsUpdater(task), task));
     } catch (RejectedExecutionException ree) {
+      // Propagate the exception to caller that has full control of the life-cycle of a helix task.
       LOGGER.error(String.format("Scheduling of task state reporter for task %s was rejected", task.getTaskId()));
+      Throwables.propagate(ree);
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java:63]: One question worth considering: Do we want to make the failure in scheduling TaskMetricsUpdater a fatal failure? or can it be made configurable?

Comment (autumnust) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java]: addressed.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java [hunk 4]
```diff
@@ -42,24 +45,39 @@
...
  * @author Yinan Li
  */
-@Alpha
+@Slf4j
 public class GobblinHelixTaskStateTracker extends AbstractTaskStateTracker {
-
-  private static final Logger LOGGER = LoggerFactory.getLogger(GobblinHelixTaskStateTracker.class);
+  @VisibleForTesting
+  static final String IS_TASK_METRICS_SCHEDULING_FAILURE_FATAl = "helixTaskTracker.isNewTaskRegFailureFatal";
+  private static final String DEFAULT_TASK_METRICS_SCHEDULING_FAILURE_FATAl = "false";
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java:52]: FATAl -> FATAL

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java [hunk 5]
```diff
@@ -42,24 +45,39 @@
  *
  * @author Yinan Li
  */
-@Alpha
+@Slf4j
 public class GobblinHelixTaskStateTracker extends AbstractTaskStateTracker {
-
-  private static final Logger LOGGER = LoggerFactory.getLogger(GobblinHelixTaskStateTracker.class);
+  @VisibleForTesting
+  static final String IS_TASK_METRICS_SCHEDULING_FAILURE_FATAl = "helixTaskTracker.isNewTaskRegFailureFatal";
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTaskStateTracker.java:51]: FATAl -> FATAL.

**Pull request closed without merging**